File: examples/extensions/php/gebext.php
<?php
const FRAME_JSON = 0;
const FRAME_BINARY = 1;

function write_frame($conn, int $type, string $payload = ''): void {
    write_all($conn, pack('NC', strlen($payload), $type) . $payload);
}

function write_all($conn, string $data): void {
    $offset = 0;
    $length = strlen($data);
    while ($offset < $length) {
        $written = fwrite($conn, substr($data, $offset));
        if ($written === false || $written === 0) {
            throw new RuntimeException('short write');
        }
        $offset += $written;
    }
}

function read_exact($conn, int $length): string {
    $data = '';
    while (strlen($data) < $length) {
        $chunk = fread($conn, $length - strlen($data));
        if ($chunk === '' || $chunk === false) {
            throw new RuntimeException('short frame');
        }
        $data .= $chunk;
    }
    return $data;
}

function read_frame($conn): ?array {
    $header = fread($conn, 5);
    if ($header === '' || $header === false) {
        return null;
    }
    while (strlen($header) < 5) {
        $header .= read_exact($conn, 5 - strlen($header));
    }
    $parts = unpack('Nlength/Ctype', $header);
    return [$parts['type'], read_exact($conn, $parts['length'])];
}

function decode_value($value, array $slots) {
    if (is_array($value) && ($value['$type'] ?? null) === 'bytes') {
        return $slots[(int)$value['slot']];
    }
    return $value;
}

function encode_value($value, array &$slots) {
    if (is_string($value) && str_starts_with($value, "bytes:")) {
        $slot = count($slots);
        $slots[] = substr($value, 6);
        return ['$type' => 'bytes', 'slot' => $slot];
    }
    return $value;
}

function serve_extension($conn, string $name, array $functions, callable $handler): void {
    sort($functions);
    write_frame($conn, FRAME_JSON, json_encode(['v' => 1, 'name' => $name, 'functions' => $functions]));
    while (true) {
        $frame = read_frame($conn);
        if ($frame === null) {
            return;
        }
        [$type, $payload] = $frame;
        if ($type !== FRAME_JSON) {
            continue;
        }
        $req = json_decode($payload, true);
        $slots = [];
        for ($i = 0; $i < (int)($req['slots'] ?? 0); $i++) {
            [$slotType, $slotPayload] = read_frame($conn);
            if ($slotType !== FRAME_BINARY) {
                throw new RuntimeException('expected binary slot');
            }
            $slots[] = $slotPayload;
        }
        if (($req['fn'] ?? '') === '__shutdown__') {
            return;
        }
        try {
            $args = array_map(fn($v) => decode_value($v, $slots), $req['args'] ?? []);
            $kwargs = [];
            foreach (($req['kwargs'] ?? []) as $key => $value) {
                $kwargs[$key] = decode_value($value, $slots);
            }
            $result = $handler($req['fn'], $args, $kwargs);
            $outSlots = [];
            $value = encode_value($result, $outSlots);
            $resp = ['id' => $req['id'], 'ok' => true, 'value' => $value];
            if ($outSlots) {
                $resp['slots'] = count($outSlots);
            }
            write_frame($conn, FRAME_JSON, json_encode($resp));
            foreach ($outSlots as $slot) {
                write_frame($conn, FRAME_BINARY, $slot);
            }
        } catch (Throwable $e) {
            write_frame($conn, FRAME_JSON, json_encode(['id' => $req['id'] ?? 0, 'ok' => false, 'error' => $e->getMessage()]));
        }
    }
}

File: examples/extensions/python/gebext.py
import json
import socket
import struct

FRAME_JSON = 0
FRAME_BINARY = 1


def write_frame(conn, frame_type, payload=b""):
    conn.sendall(struct.pack(">IB", len(payload), frame_type) + payload)


def read_frame(conn):
    header = conn.recv(5)
    if not header:
        return None, None
    while len(header) < 5:
        chunk = conn.recv(5 - len(header))
        if not chunk:
            raise EOFError("short frame header")
        header += chunk
    length, frame_type = struct.unpack(">IB", header)
    payload = b""
    while len(payload) < length:
        chunk = conn.recv(length - len(payload))
        if not chunk:
            raise EOFError("short frame payload")
        payload += chunk
    return frame_type, payload


def encode_value(value, slots):
    if isinstance(value, bytes):
        slot = len(slots)
        slots.append(value)
        return {"$type": "bytes", "slot": slot}
    return value


def decode_value(value, slots):
    if isinstance(value, dict) and value.get("$type") == "bytes":
        return slots[int(value["slot"])]
    return value


def serve_connection(conn, name, functions, handler):
    write_frame(conn, FRAME_JSON, json.dumps({
        "v": 1,
        "name": name,
        "functions": sorted(functions),
    }).encode())
    while True:
        frame_type, payload = read_frame(conn)
        if frame_type is None:
            return
        if frame_type != FRAME_JSON:
            continue
        req = json.loads(payload.decode())
        slots = []
        for _ in range(int(req.get("slots", 0))):
            slot_type, slot_payload = read_frame(conn)
            if slot_type != FRAME_BINARY:
                raise ValueError("expected binary slot")
            slots.append(slot_payload)
        if req.get("fn") == "__shutdown__":
            return
        try:
            args = [decode_value(arg, slots) for arg in req.get("args", [])]
            kwargs = {k: decode_value(v, slots) for k, v in req.get("kwargs", {}).items()}
            result = handler(req["fn"], args, kwargs)
            out_slots = []
            value = encode_value(result, out_slots)
            resp = {"id": req["id"], "ok": True, "value": value}
            if out_slots:
                resp["slots"] = len(out_slots)
            write_frame(conn, FRAME_JSON, json.dumps(resp).encode())
            for slot in out_slots:
                write_frame(conn, FRAME_BINARY, slot)
        except Exception as exc:
            write_frame(conn, FRAME_JSON, json.dumps({
                "id": req.get("id", 0),
                "ok": False,
                "error": str(exc),
            }).encode())


def listen(sock_path=None, host="127.0.0.1", port=9101):
    if sock_path:
        server = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        server.bind(sock_path)
    else:
        server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server.bind((host, port))
    server.listen(1)
    return server
